docs/tutorials/qml/ml_tools/trainer.md

```diff
@@ -0,0 +1,615 @@
+Default is gradient-based optimization.
+
+- **Gradient-Based Optimization**: Utilizes optimizers from PyTorch's `torch.optim` module.
+This is the default behaviour of the `Trainer`, thus setting this is not necessary.
+However, it can be explicity mentioned as follows.
+Example of using gradient-based optimization:
+
+```python exec="on" source="material-block"
+from qadence.ml_tools import Trainer
+
+# set_use_grad(True) to enable gradient based training. This is the default behaviour of Trainer.
+Trainer.set_use_grad(True)
+```
+
+- **Gradient-Free Optimization**: Employs optimization algorithms from the [Nevergrad](https://facebookresearch.github.io/nevergrad/) library.
+
+
+Example of using gradient-free optimization with Nevergrad:
+
+```python exec="on" source="material-block"
+from qadence.ml_tools import Trainer
+
+# set_use_grad(False) to disable gradient based training.
+Trainer.set_use_grad(False)
+```
+
+### Using Context Managers for Mixed Optimization
+
+For cases requiring both optimization methods in a single training session, the `Trainer` class provides context managers to enable or disable gradients.
+
+```python
+# Temporarily switch to gradient-based optimization
+with trainer.enable_grad_opt(optimizer):
+    print("Gradient Based Optimization")
+    # trainer.fit(train_loader)
+
+# Switch to gradient-free optimization for specific steps
+with trainer.disable_grad_opt(ng_optimizer):
+    print("Gradient Free Optimization")
+    # trainer.fit(train_loader)
+```
+
+---
+
+## 3. Custom Loss Functions
+
+Users can define custom loss functions tailored to their specific tasks.
+The `Trainer` accepts a `loss_fn` parameter, which should be a callable that takes the model and data as inputs and returns a tuple containing the loss tensor and a dictionary of metrics.
+
+Example of using a custom loss function:
+
+```python exec="on" source="material-block"
+import torch
+from itertools import count
+cnt = count()
+criterion = torch.nn.MSELoss()
+
+def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, dict]:
+    next(cnt)
+    x, y = data
+    out = model(x)
+    loss = criterion(out, y)
+    return loss, {}
+```
+
+---
+
+## 4. Hooks for Custom Behavior
+
+The `Trainer` class provides several hooks that enable users to inject custom behavior at different stages of the training process.
+These hooks are methods that can be overridden in a subclass to execute custom code.
+The available hooks include:
+
+- `on_train_start`: Called at the beginning of the training process.
+- `on_train_end`: Called at the end of the training process.
+- `on_train_epoch_start`: Called at the start of each training epoch.
+- `on_train_epoch_end`: Called at the end of each training epoch.
+- `on_train_batch_start`: Called at the start of each training batch.
+- `on_train_batch_end`: Called at the end of each training batch.
+
+Example of using a hook to log a message at the end of each epoch:
```

Comment:
Indeed. 
Callbacks are mainly for logging/tracking purposes - but these hooks are generic. 
Eg. In every train batch start, 
1. perform on_train_batch_start callbacks (using the decorator BaseTrainer.callback("train_batch"))
2. call the on_train_batch_start hook
3. do the batch training
4. call the on_train_batch_end hook
5. perform on_train_batch_end callbacks (using the decorator BaseTrainer.callback("train_batch"))

The use of these hooks is not specifically to add extra callbacks, but for any other generic pre/post processing 
1. eg, reshaping input batch in case of RNNs/LSTMs, post processing loss and adding an extra metric. 
2. they could also be used to add more callbacks (which is not recommended - as we provide methods to add extra callbacks in the TrainCofig)

Reply:
Right, thanks for the explanation. Is the difference clear from the docs? Maybe I missed it.